src/physics/em/detail/UrbanMscScatter.hh

```diff
@@ -505,29 +469,73 @@ CELER_FUNCTION real_type UrbanMscScatter::calc_correction(real_type tau) const
 //---------------------------------------------------------------------------//
 /*!
  * Sample the displacement direction using G4UrbanMscModel::SampleDisplacement
- * of Geant4 10.7: simple and fast sampling based on single scattering results.
+ * (simple and fast sampling based on single scattering results) and update
+ * direction and position of the particle.
  *
  * A simple distribution for the unit direction on the lateral (x-y) plane,
  * \f$ Phi = \phi \pm \psi \f$ where \f$ psi \sim \exp(-\beta*v) \f$ and
  * \f$\beta\f$ is determined from the requirement that the distribution should
  * give the same mean value that is obtained from the single scattering
  * simulation.
  *
- * @param phi the azimuthal angle of the multiple scattering.
+ * @param cth the cos(\theta) of the multiple scattering.
+ * @param rmax2 the asymmetry between true path and geom path
  */
 template<class Engine>
-CELER_FUNCTION Real3 UrbanMscScatter::dir_displacement(Engine&   rng,
-                                                       real_type phi) const
+CELER_FUNCTION void
+UrbanMscScatter::update_geometry(Engine& rng, real_type cth, real_type rmax2)
 {
-    constexpr real_type cbeta  = 2.160;
-    real_type           cbeta1 = 1 - std::exp(-cbeta * constants::pi);
+    // Update direction
+    UniformRealDistribution<real_type> sample_phi(0, 2 * constants::pi);
+    real_type                          phi = sample_phi(rng);
+    Real3 direction = rotate(from_spherical(cth, phi), inc_direction_);
+
+    geometry_.set_dir(direction);
+
+    // Displace the lateral position and update position
+    if (input_.is_displaced && tau_ >= params_.tau_small && rmax2 > 0)
+    {
+        // Sample a unit direction of the displacement
+        constexpr real_type cbeta  = 2.160;
+        real_type           cbeta1 = 1 - std::exp(-cbeta * constants::pi);
+
+        real_type psi = -std::log(1 - generate_canonical(rng) * cbeta1) / cbeta;
+        real_type angle = BernoulliDistribution(0.5)(rng) ? phi + psi
+                                                          : phi - psi;
 
-    real_type psi   = -std::log(1 - generate_canonical(rng) * cbeta1) / cbeta;
-    real_type angle = BernoulliDistribution(0.5)(rng) ? phi + psi : phi - psi;
+        Real3 displacement{std::cos(angle), std::sin(angle), 0};
+
+        // Rotate along the incident particle direction
+        displacement = rotate(displacement, inc_direction_);
+
+        // Scale with the lateral arm
+        real_type arm = real_type(0.73) * std::sqrt(rmax2);
+        for (auto i : range(3))
+        {
+            displacement[i] *= arm;
+        }
 
-    Real3 dir{std::cos(angle), std::sin(angle), 0};
+        real_type rho = norm(displacement);
 
-    return dir;
+        if (rho > params_.geom_limit)
+        {
+            // Update position
+            Real3     new_pos = geometry_.pos();
+            real_type safety  = real_type(0.99)
```

Comment:
Why the arbitrary value? Might be good to add a
```c++
static constexpr real_type safety_tol() { return 0.01; }
```
and change this to
```suggestion
            real_type safety  = (1 - safety_tol())
```

Reply:
Added safety_tol in UrbanMscParameters.